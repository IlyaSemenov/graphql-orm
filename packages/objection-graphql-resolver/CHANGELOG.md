# objection-graphql-resolver

## 7.0.0

### Major Changes

- 463ebee: Merge objection-graphql-resolver and orchid-orm, extract common core to graphql-orm-resolver.
