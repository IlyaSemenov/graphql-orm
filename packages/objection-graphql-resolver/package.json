{
  "name": "objection-graphql-resolver",
  "description": "objection.js plugin: resolve GraphQL query using withGraphFetched",
  "version": "7.0.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/IlyaSemenov/objection-graphql-resolver.git"
  },
  "author": "Ilya Semenov",
  "license": "MIT",
  "type": "module",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "source": "src/index.ts",
  "files": [
    "dist",
    "src"
  ],
  "engines": {
    "node": ">=14"
  },
  "scripts": {
    "dev": "vite-node --watch playground/readme.sample.ts",
    "build": "tsup",
    "prepack": "npm run build",
    "test": "vitest run"
  },
  "dependencies": {
    "graphql-parse-resolve-info": "^4.12.0"
  },
  "peerDependencies": {
    "graphql": "^16",
    "objection": "^2 || ^3"
  },
  "devDependencies": {
    "@apollo/server": "~4.7.1",
    "@tsconfig/node14": "~1.0.3",
    "graphql": "~16.6.0",
    "graphql-request": "~6.0.0",
    "graphql-tag": "^2.12.6",
    "knex": "~2.4.2",
    "objection": "^3.0.1",
    "sqlite3": "~5.1.6",
    "tar": "~6.1.15",
    "tsup": "~6.7.0",
    "typescript": "~5.0.4",
    "vite": "~4.3.8",
    "vite-node": "~0.31.1",
    "vite-tsconfig-paths": "~4.2.0",
    "vitest": "~0.31.1"
  }
}
