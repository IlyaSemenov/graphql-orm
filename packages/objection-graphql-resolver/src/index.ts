export {
	createGraphResolver,
	createGraphResolver as graph,
} from "./resolvers/graph"
export {
	defineModelResolver,
	defineModelResolver as model,
} from "./resolvers/model"
export {
	defineCursorPaginator as cursor,
	defineCursorPaginator,
} from "graphql-orm-resolver"
export {
	defineFieldResolver,
	defineFieldResolver as field,
} from "graphql-orm-resolver"
export {
	definePageResolver,
	definePageResolver as page,
} from "graphql-orm-resolver"
export {
	defineRelationResolver,
	defineRelationResolver as relation,
} from "graphql-orm-resolver"
