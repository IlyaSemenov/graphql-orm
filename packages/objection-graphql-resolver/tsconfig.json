{
	"extends": "@tsconfig/node14/tsconfig.json",
	"compilerOptions": {
		"module": "ESNext",
		"paths": {
			"graphql-orm": ["../graphql-orm/src"],
			"objection-graphql-resolver": ["./src"]
		}
	}
}
