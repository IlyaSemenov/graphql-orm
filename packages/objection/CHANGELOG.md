# objection-graphql-resolver

## 7.4.0

### Minor Changes

- f3f5f31: Type graph resolve result as `Promise<T>` instead of `Query`.

### Patch Changes

- e802bea: Return `Promise<undefined>` instead of `undefined` for non-requested fields.

## 7.3.1

### Patch Changes

- 05d7316: Revert inlining graphql-parse-resolve-info, bump its version instead.

## 7.3.0

### Minor Changes

- 262aa3e: Inline graphql-parse-resolve-info to prevent false peer dependency error in dependent packages.

## 7.2.2

### Patch Changes

- 2f28c2b: Support `modelField` in addition to `tableField` as per docs (fix #33).

## 7.2.1

### Patch Changes

- 2aa2a48: Fix root subfield lookup when the field was not requested by the client.

## 7.2.0

### Minor Changes

- 49ea582: Support resolving non-root GraphQL fields (#27).

### Patch Changes

- b10f1d2: Make GraphResolveOptions a <Context> generic.

## 7.1.0

### Minor Changes

- 030b302: Define graph resolver context type and pass it to callbacks.

## 7.0.0

### Major Changes

- 463ebee: Merge objection-graphql-resolver and orchid-orm, extract common core to graphql-orm.
