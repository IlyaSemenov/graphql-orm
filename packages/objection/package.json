{
  "name": "objection-graphql-resolver",
  "type": "module",
  "version": "7.6.0",
  "description": "objection.js plugin: resolve GraphQL query using withGraphFetched",
  "author": "Ilya Semenov",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/IlyaSemenov/objection-graphql-resolver.git"
  },
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    }
  },
  "main": "dist/index.cjs",
  "files": [
    "dist",
    "src"
  ],
  "engines": {
    "node": ">=14"
  },
  "scripts": {
    "dev": "vite-node --watch playground/readme.sample.ts",
    "build": "tsup",
    "test": "vitest run && tsc --noEmit",
    "prepublishOnly": "pnpm build"
  },
  "peerDependencies": {
    "graphql": "^16",
    "objection": "^2 || ^3"
  },
  "dependencies": {
    "graphql-parse-resolve-info": "^4.13.0"
  },
  "devDependencies": {
    "@apollo/server": "^4.7.1",
    "graphql": "^16.6.0",
    "graphql-request": "^6.0.0",
    "graphql-tag": "^2.12.6",
    "knex": "^2.4.2",
    "objection": "^3.0.1",
    "sqlite3": "^5.1.6",
    "tar": "^6.1.15",
    "tsconfig-vite-node": "^1.1.2",
    "tsup": "^8.5.0",
    "vite": "^7.0.5",
    "vite-node": "^3.2.4",
    "vite-tsconfig-paths": "^5.1.4",
    "vitest": "^3.2.4"
  }
}
