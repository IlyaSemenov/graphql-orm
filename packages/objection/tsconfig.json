{
	"extends": "tsconfig-vite-node/tsconfig.json",
	"compilerOptions": {
		// Using "target": "ESNext" (default coming from tsconfig-vite-node) breaks sqlite.
		// tests/main/m2m.test.ts: SQLITE_CONSTRAINT: NOT NULL constraint failed
		"target": "ES2021",
		"paths": {
			"objection-graphql-resolver": ["./src"],
			"graphql-orm": ["../base/src"]
		}
	}
}
