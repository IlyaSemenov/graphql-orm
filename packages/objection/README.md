# objection-graphql-resolver

A helper library to resolve GraphQL queries directly with [Objection.js](https://vincit.github.io/objection.js/) models and relations.

- Highly effective: selects only requested fields and relations (using fine-tuned `withGraphFetched`).
- Supports unlimited nested resolvers (traversing `relationMappings`).
- Supports pagination.
- Supports virtual attributes.
- Supports filters like `{ date: "2020-10-01", category__in: ["News", "Politics"] }`.
- Hooks into subqueries with query modifiers.
- Hooks into field results to restrict access to sensitive information.

_Note that there is a sister project: `orchid-graphql` which does the same for [Orchid ORM](https://orchid-orm.netlify.app)._

## Install

```sh
npm i objection-graphql-resolver
```

## Minimal all-in-one example

Run GraphQL server:

```ts
// Everything is put into a single file for demonstration purposes.
//
// In real projects, you will want to separate models, typedefs,
// model resolvers, and the server into their own modules.

import { ApolloServer, ApolloServerOptions } from "@apollo/server"
import { startStandaloneServer } from "@apollo/server/standalone"
import gql from "graphql-tag"
import Knex from "knex"
import { Model } from "objection"
import * as r from "objection-graphql-resolver"

// Define Objection.js models

class PostModel extends Model {
  static tableName = "post"

  id?: number
  text?: string
}

// Define GraphQL schema

const typeDefs = gql`
  type Post {
    id: Int!
    text: String!
  }

  type Mutation {
    create_post(text: String!): Post!
  }

  type Query {
    posts: [Post!]!
  }
`

// Map GraphQL types to model resolvers

const graph = r.graph({
  Post: r.model(PostModel),
})

// Define resolvers

const resolvers: ApolloServerOptions<any>["resolvers"] = {
  Mutation: {
    async create_post(_parent, args, context, info) {
      const post = await PostModel.query().insert(args)
      return graph.resolve(post.$query(), { context, info })
    },
  },
  Query: {
    posts(_parent, _args, context, info) {
      return graph.resolve(PostModel.query().orderBy("id"), { context, info })
    },
  },
}

// Configure database backend

const knex = Knex({ client: "sqlite3", connection: ":memory:" })
Model.knex(knex)

await knex.schema.createTable("post", (post) => {
  post.increments("id")
  post.text("text").notNullable()
})

// Start GraphQL server

const server = new ApolloServer({ typeDefs, resolvers })
const { url } = await startStandaloneServer(server, {
  listen: { port: 4000 },
})
console.log(`Listening on ${url}`)
```

Query it with GraphQL client:

```ts
import { GraphQLClient } from "graphql-request"
import gql from "graphql-tag"

const client = new GraphQLClient("http://127.0.0.1:4000")

await client.request(
  gql`
    mutation create_post($text: String!) {
      new_post: create_post(text: $text) {
        id
      }
    }
  `,
  { text: "Hello, world!" },
)

const { posts } = await client.request(gql`
  query {
    posts {
      id
      text
    }
  }
`)

console.log(posts)
```

## Relations

Relations will be fetched automatically using `withGraphFetched()` when resolving nested fields.

Example:

```ts
const graph = r.graph({
  User: r.model(UserModel),
  Post: r.model(PostModel),
})
```

```gql
query posts_with_author {
  posts {
    id
    text
    # will use withGraphFetched("author") if requested
    author {
      name
    }
  }
}

query user_with_posts($id: ID!) {
  user(id: $id) {
    name
    # will use withGraphFetched("posts") if requested
    posts {
      id
      text
    }
  }
}
```

[More details and examples for relations.](docs/relations.md)

## Fields access

Access to individual fields can be limited:

```ts
const graph = r.graph({
  User: r.model(UserModel, {
    fields: {
      id: true,
      name: true,
      // other fields not specified here, such as user password,
      // will not be accessible
    },
  }),
})
```

This API also allows to fine-tune field selectors, see [API](#api) section below.

## Pagination

Root queries and -to-many nested relations can be paginated.

```ts
const graph = r.graph({
  User: r.model(UserModel, {
    fields: {
      id: true,
      name: true,
      // user.posts will be a page with nodes and continuation cursor
      posts: r.page(r.cursor({ fields: ["-id"], take: 10 })),
    },
  }),
  Post: r.model(PostModel),
})
```

To paginate root query, use:

```ts
const resolvers = {
  Query: {
    posts: async (parent, args, context, info) => {
      return graph.resolvePage(
        PostModel.query(),
        r.cursor({ take: 10, fields: ["-id"] }),
        { context, info },
      )
    },
  },
}
```

[More details and examples for pagination.](docs/pagination.md)

## Filters

Both root and nested queries can be filtered with GraphQL arguments:

```gql
query {
  posts(filter: { date: "2020-10-01", author_id__in: [123, 456] }) {
    id
    date
    text
    author {
      id
      name
    }
  }
}
```

Filters will run against database fields, or call model modifiers.

[More details and examples for filters.](docs/filters.md)

## Virtual attributes

Virtual attributes (getters on models) can be accessed the same way as database fields:

```ts
class PostModel extends Model {
  declare id?: number
  declare title?: string

  get url() {
    assert(this.id)
    return `/${this.id}.html`
  }
}
```

```gql
query {
  posts {
    id
    title
    url
  }
}
```

[More details and examples for virtual attributes.](docs/virtual-attributes.md)

## API

```ts
import * as r from "objection-graphql-resolver"

const graph = r.graph(
  // Map GraphQL types to model resolvers (required)
  {
    Post: r.model(
      // Objection.js model class (required)
      PostModel,
      // Model resolver options
      {
        // List fields that can be accessed via GraphQL
        // if not provided, all fields can be accessed
        fields: {
          // Select field from database
          id: true,
          // Call model getter with this name
          url: true,
          // Descend into relation
          // (related model must be also registered in this graph resolver)
          author: true,
          // Modify query when this field is resolved
          preview: query =>
            query.select(raw("substr(text,1,100) as preview")),
          // Same as text: true
          text: r.field(),
          // Custom field resolver
          text2: r.field({
            // Model (database) field, if different from GraphQL field
            modelField: "text",
          }),
          preview2: r.field({
            // Modify query
            modify: query =>
              query.select(raw("substr(text,1,100) as preview2")),
            // Post-process selected value
            transform(
              // Selected value
              preview,
              // Current instance
              post,
              // Field resolve context: graph, tree, type, field, filters, context
              context,
            ) {
              if (preview.length < 100) {
                return preview
              } else {
                return preview + "..."
              }
            },
          }),
          // Select all objects in -to-many relation
          comments: true,
          // Select all objects in -to-many relation
          all_comments: r.relation({
            // Model field, if different from GraphQL field
            modelField: "comments",
            // Enable filters on -to-many relation
            filters: true,
            // Modify subquery
            modify: (query, { liked }) =>
              query.where({ liked }).orderBy("id", "desc"),
            // Post-process selected values, see r.field()
            // transform: ...,
          }),
          // Paginate subquery in -to-many relation
          comments_page: r.page(
            // Paginator
            r.cursor(
              // Pagination options
              // Default: { fields: ["id"], take: 10 }
              {
                // Which fields to use for ordering
                // Prefix with - for descending sort
                fields: ["name", "-id"],
                // How many object to take per page
                take: 10,
              },
            ),
            {
              // All r.relation() options, such as:
              modelField: "comments",
            },
          ),
        },
        // Modify all queries to this model
        modify: (
          // ORM (sub)query
          query,
          // Table resolve context: graph, tree, type, filters, context
          context,
        ) => query.where(context.tree.args).orderBy("id", "desc"),
        // Allow all fields (`fields` will be used for overrides)
        allowAllFields: true,
        // Allow filters in all relations
        allowAllFilters: true,
      },
    ),
  },
  // Graph options
  {
    // Allow all fields in all models (`fields` will be used for overrides)
    allowAllFields: true,
    // Allow filters in all models' relations
    allowAllFilters: true,
  },
)

const resolvers = {
  Query: {
    posts: (parent, args, context, info) => {
      return graph.resolve(
        // Root query (required)
        PostModel.query(),
        // Options (required)
        {
          // Resolver context
          context,
          // GraphQLResolveInfo object, as passed by GraphQL executor (required)
          info,
          // Enable filters
          filters: true,
        },
      )
    },
    posts_page: (parent, args, context, info) => {
      return graph.resolvePage(
        // Root query (required)
        PostModel.query(),
        // Paginator (required)
        r.cursor({ fields: ["-id"], take: 10 }),
        // Options (required) - see graph.resolve
        {
          context,
          info,
          filters: true,
        },
      )
    },
  },
}

graph.resolve(PostModel.query().findById(1), {
  // As comes from Apollo
  context,
  // As comes from Apollo
  info,
  // If resolving subfield
  path: ["subfield"],
})
```
