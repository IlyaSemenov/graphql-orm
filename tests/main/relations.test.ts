import gql from "graphql-tag"
import { Model } from "objection"
import * as r from "objection-graphql-resolver"
import { expect, test } from "vitest"

import { Resolvers, setup } from "../setup"

class UserModel extends Model {
	static tableName = "user"

	static get relationMappings() {
		return {
			posts: {
				relation: Model.HasManyRelation,
				modelClass: PostModel,
				join: { from: "user.id", to: "post.author_id" },
			},
		}
	}

	id?: number
	name?: string
	posts?: PostModel[]
}

class PostModel extends Model {
	static tableName = "post"

	static get relationMappings() {
		return {
			author: {
				relation: Model.BelongsToOneRelation,
				modelClass: UserModel,
				join: { from: "post.author_id", to: "user.id" },
			},
		}
	}

	id?: number
	text?: string
	author_id?: number
	author?: UserModel
}

const schema = gql`
	scalar Filter

	type User {
		id: Int!
		name: String!
		posts: [Post!]!
	}

	type Post {
		id: Int!
		text: String!
		author: User!
	}

	type Query {
		posts: [Post!]!
	}
`

const graph = r.graph(
	{
		User: r.model(UserModel),
		Post: r.model(PostModel),
	},
	{
		allowAllFields: true,
	}
)

const resolvers: Resolvers = {
	Query: {
		posts(_parent, _args, ctx, info) {
			return graph.resolve(ctx, info, PostModel.query())
		},
	},
}

const { client, knex } = await setup({ typeDefs: schema, resolvers })

test("filters", async () => {
	await knex.schema.createTable("user", (user) => {
		user.increments("id").notNullable().primary()
		user.string("name").notNullable()
	})

	await knex.schema.createTable("post", (post) => {
		post.increments("id").notNullable().primary()
		post.string("text").notNullable()
		post.integer("author_id").notNullable().references("user.id")
	})

	await UserModel.query().insertGraph([
		{ name: "Alice" },
		{ name: "Bob" },
		{ name: "Charlie" },
	])

	await PostModel.query().insertGraph(
		[
			{ author_id: 1, text: "Oil price rising." },
			{ author_id: 1, text: "Is communism dead yet?" },
			{ author_id: 2, text: "Latest COVID news." },
		],
		{ relate: true }
	)

	expect(
		await client.request(
			gql`
				{
					posts {
						id
						text
						author {
							id
							name
						}
					}
				}
			`
		)
	).toMatchSnapshot()

	expect(
		await client.request(
			gql`
				{
					posts {
						id
						text
						author {
							id
							name
							posts {
								id
								text
							}
						}
					}
				}
			`
		)
	).toMatchSnapshot()
})
