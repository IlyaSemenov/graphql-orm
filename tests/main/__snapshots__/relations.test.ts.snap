// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filters 1`] = `
{
  "posts": [
    {
      "author": {
        "id": 1,
        "name": "Alice",
      },
      "id": 1,
      "text": "Oil price rising.",
    },
    {
      "author": {
        "id": 1,
        "name": "Alice",
      },
      "id": 2,
      "text": "Is communism dead yet?",
    },
    {
      "author": {
        "id": 2,
        "name": "Bob",
      },
      "id": 3,
      "text": "Latest COVID news.",
    },
  ],
}
`;

exports[`filters 2`] = `
{
  "posts": [
    {
      "author": {
        "id": 1,
        "name": "Alice",
        "posts": [
          {
            "id": 1,
            "text": "Oil price rising.",
          },
          {
            "id": 2,
            "text": "Is communism dead yet?",
          },
        ],
      },
      "id": 1,
      "text": "Oil price rising.",
    },
    {
      "author": {
        "id": 1,
        "name": "Alice",
        "posts": [
          {
            "id": 1,
            "text": "Oil price rising.",
          },
          {
            "id": 2,
            "text": "Is communism dead yet?",
          },
        ],
      },
      "id": 2,
      "text": "Is communism dead yet?",
    },
    {
      "author": {
        "id": 2,
        "name": "Bob",
        "posts": [
          {
            "id": 3,
            "text": "Latest COVID news.",
          },
        ],
      },
      "id": 3,
      "text": "Latest COVID news.",
    },
  ],
}
`;
